src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -249,11 +346,13 @@ public void Initialize()
             {
                 _deadline = Clock.UtcNow.Add(timeout);
 
-                // Set timer field first and then set the timeout.
-                // The method uses the timer to lock and we want to avoid the timer immediately
-                // running before the field is set.
-                _deadlineTimer = new Timer(DeadlineExceeded);
-                _deadlineTimer.Change(timeout, Timeout.InfiniteTimeSpan);
+                // Create lock before setting up deadline event
+                _deadlineLock = new SemaphoreSlim(1, 1);
+
+                _deadlineCts = new CancellationTokenSource(timeout);
+                _deadlineCts.Token.Register(DeadlineExceeded);
+
+                _requestAbortedRegistration = HttpContext.RequestAborted.Register(RequestAborted);
```

Comment:
Rather than register against RequestAborted here, I'd expected you to call CTS.CreateLinkedTokenSource to combine _deadlineCts and RequestAborted and then replace HttpContext.RequestAborted with the combine token. That way the app would be notified if the deadline expired.

Reply:
What do you mean by the app? ServerCallContext exposes its own token which is what users should use in their gRPC services

https://github.com/grpc/grpc-dotnet/blob/f721fbef75e485b252588002da4c7b919666f9c8/src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs#L202

Reply:
It should be possible to use CreateLinkedTokenSource only when there are at least two cancellable source tokens and otherwise save the allocations, but it probably ends up not looking any better than what you already have.

E.g. https://github.com/aspnet/AspNetCore/blob/master/src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs#L1502-L1520

Reply:
CancellationTokenCore gives them the result of one token or the other, not both. E.g. if you're using deadlines then the app doesn't know about RequestAborted. That should be a linked source so they can can be triggered by either.

Reply:
> CancellationTokenCore gives them the result of one token or the other, not both. E.g. if you're using deadlines then the app doesn't know about RequestAborted. That should be a linked source so they can can be triggered by either.

If there is no deadline then only HttpContext.RequestAborted can raise the event.

If there is a deadline then _deadlineCts gets raised when HttpContext.RequestAborted is raised.

I'll investigate replacing the registration with a linked cancellation token but I suspect it will allocate more and not improve code readability much.